clash-cores/src/Clash/Cores/Xilinx/DcFifo/Explicit.hs

```diff
@@ -0,0 +1,246 @@
+  { hsQueue      :: Seq.Seq (BitVector n)
+  , relativeTime :: Int
+  } deriving Show
+
+data XilinxFifo read write depth n =
+  XilinxFifo
+    { writeReset :: Signal write ResetBusy
+    , isFull :: Signal write Full
+    , isOverflow :: Signal write Bool
+    , writeCount :: Signal write (DataCount depth)
+    , readReset :: Signal read ResetBusy
+    , isEmpty :: Signal read Empty
+    , isUnderflow :: Signal read Bool
+    , readCount :: Signal read (DataCount depth)
+    , fifoData :: Signal read (BitVector n)
+    }
+
+-- | Mock Xilinx FIFO, see [documentation](https://docs.xilinx.com/v/u/en-US/pg057-fifo-generator)
+-- for expected behavior.
+--
+-- If 'dcReadDataCount' or 'dcWriteDataCount' are disabled, the relevant signals
+-- will be 'deepErrorX'
+dcFifo ::
+  forall depth n write read .
+  ( KnownNat n
+  , KnownDomain write
+  , KnownDomain read
+
+  , KnownNat depth
+  -- Number of elements should be between [2**4, 2**17] ~ [16, 131072].
+  , 4 <= depth
+  , depth <= 17
+  ) =>
+  DcConfig (SNat depth) ->
+
+  Clock write -> Clock read ->
+  -- | Asynchronous reset
+  Reset read ->
+
+  -- | Write data
+  Signal write (Maybe (BitVector n)) ->
+  -- | Read enable
+  Signal read Bool ->
+  XilinxFifo read write depth n
+dcFifo DcConfig{..} wClk rClk rst writeData rEnable =
+  let
+    (wRstBusy, wFull, wOver, wCnt, rRstBusy, rEmpty, rUnder, rCnt, rData) =
+      go initState rstSignalR rEnable rstSignalW writeData
+  in XilinxFifo
+      wRstBusy
+      wFull
+      (if dcOverflow then wOver else deepErrorX "Overflow disabled")
+      (if dcWriteDataCount then wCnt else deepErrorX "Write data count disabled")
+      rRstBusy
+      rEmpty
+      (if dcUnderflow then rUnder else deepErrorX "Underflow disabled")
+      (if dcReadDataCount then rCnt else deepErrorX "Read data count disabled")
+      (deepErrorX "No sample" :- rData)
+
+ where
+  rstSignalR = unsafeToHighPolarity rst
+  rstSignalW = unsafeSynchronizer rClk wClk $ unsafeToHighPolarity rst
+
+  -- reified depth
+  maxDepth = snatToNum @Int (powSNat (SNat @2) (SNat @depth))
+
+  -- TODO: how do we ensure it doesn't "space leak" in Haskell?
+  --
+  -- force slower first?
+  -- https://github.com/clash-lang/clash-compiler/blob/ea114d8edd6a110f72d148203b9db2454cae8f37/clash-prelude/src/Clash/Explicit/BlockRam.hs#L1276-L1280
+
+  go ::
+    FifoState n ->
+    Signal read Bool -> -- reset
+    Signal read Bool -> -- read enabled
+    Signal write Bool -> -- reset
+    Signal write (Maybe (BitVector n)) -> -- write data
+    ( Signal write ResetBusy
+    , Signal write Full
+    , Signal write Bool
+    , Signal write (DataCount depth)
+
+    , Signal read ResetBusy
+    , Signal read Empty
+    , Signal read Bool
+    , Signal read (DataCount depth)
+    , Signal read (BitVector n)
+    )
+  go st@(FifoState _ rt) rstR rEna rstW =
+    if rt < tWr
+      then goRead st rstR rEna rstW
+      else goWrite st rstR rEna rstW
+    -- TODO: goBoth case?
```

Comment:
I think we can let this slide, unless we want a 1:1 model to the behavioral RTL. But I doubt we want that, given that the behavioral model doesn't correspond to the generated hardware anyway.